### ISSUE TITLE
Github OAuth times out

### ISSUE THREAD

**Issue by pierrebeaucamp** [concourse/concourse#973]
Hi,

I'm trying to set up a team using Github OAuth. When trying to log in, we get successfully redirected to Github, where we need to log in / grant our app access to Github. Once this is done, we get redirected back to concourse, with the following call:

Request URL: `http://concourse.karma.io/auth/github/callback?code=45xxxxx78&state=eyxxxxxn0`
Referer: `http://concourse.karma.io/teams/ilovetravel/login`
Cookie: `_concourse_oauth_state=eyxxxxxxn0 (same state)`

This call just times out after a while. It appears to do nothing at all.

Same thing happens if I try to log in over the fly cli: The link redirects me to github, I authenticate and then I just wait indefinitely. Interestingly enough, the cli link contains `fly_local_port=59552`. Do I need to whitelist additional ports in our firewall? 

* Concourse version: 2.7.0
* Deployment type (BOSH/Docker/binary): BOSH
* Infrastructure/IaaS: AWS
* Browser (if applicable): Tested on both Chrome & Firefox
* Did this used to work? No

**Comment by vito** [comment]
Is your external URL configured correctly, and reachable from your workstation? That request should be initiated from your machine.

**Comment by pierrebeaucamp** [comment]
yes, the external URL seems to be configured correctly. Also, if I try to visit to the `Authorization callback URL` from the settings page in Github, it resolves correctly to our concourse instance.

Is there maybe an easy way to access the access log of concourse? That way I can at least check if the request makes to our instance.

**Comment by vito** [comment]
Make sure your ATC can reach your GitHub instance. It needs to call back out to it in order to verify your user's access.

**Comment by pierrebeaucamp** [comment]
Our ATC can reach GitHub just fine. Further, I temporarily disabled all of our firewall rules to make sure we're not blocking this request - however that didn't help either.

**Comment by richarddowner** [comment]
Also having this problem

**Comment by pierrebeaucamp** [comment]
@richarddowner I'm glad that I'm not the only one having this problem. Unfortunately, we still couldn't resolve that from our end (We're currently running 3.0.1). I start to think that it might be a problem with our bosh, but I didn't have the time to redeploy a new bosh to see if the problem persists.

**Comment by mariash** [comment]
@pierrebeaucamp can you check ATC logs if there is anything interesting there when callback happens. Logs will be located at `/var/vcap/sys/log/atc`. Make sure your ATC `log_level` is set to `debug`.

Also during callback try to get the dump of goroutines in ATC by curling `localhost:8079/debug/pprof/goroutine?debug=2` on ATC VM. Search for the ones in `auth`. Hopefully, it will give you the line that is hanging.

**Comment by richarddowner** [comment]
@pierrebeaucamp our concourse instance is within a private aws vpc and the DNS name is not resolvable outside certain networks. We have decided the problem must be something to do with public github not being able to resolve our concourse instance's DNS name on an auth callback somewhere behind the scenes.

**Comment by mariash** [comment]
@richarddowner yes that sounds like that will cause this. I am going to close this issue then.

**Comment by richarddowner** [comment]
Hey @mariash and @pierrebeaucamp - There has been a new development with this. I think ive narrowed down the cause. And its not what I previously thought. 

I believe either concourse web OR the golang oauth package is not honoring the http_proxy, https_proxy and no_proxy environment variables. 

I managed to get this working in AWS by adding a transparent proxy between our corporate proxy and the concourse instance. This proxy adds a default route 0.0.0.0/0 to the subnet route table so ALL traffic is routed to the transparent proxy and then on to corporate proxy. The oAuth now works. 

Here are some of my findings:

**Error within concourse atc:**
```
atc.oauth-callback.callback.failed-to-exchange-token, errror: post https://github.mycompany.com/login/oauth/access_token: dial tcp xx.xx.xx.xx:443: getsockopt: connection timed out
```

**The call within concourse throwing the error:**
https://github.com/concourse/atc/blob/176c130dc78f77eb0350b42c8b40312a3cbf3547/auth/oauth_callback_handler.go#L147

**The library that concourse calls for oauth exchange:**
https://github.com/golang/oauth2/blob/4464e7848382e6f39db55097f70e1460bdf944b3/oauth2.go#L178

I cannot confirm, but it seems like, either concourse atc or the golang oauth2 module is not honoring the proxy environment variables.

I am going to open this as a new issue.
